pkg/volume/csi/csi_client.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/golang/glog"
+	grpctx "golang.org/x/net/context"
+	"google.golang.org/grpc"
+	api "k8s.io/api/core/v1"
+	csipb "k8s.io/kubernetes/pkg/volume/csi/proto/csi"
+)
+
+type csiClient interface {
+	AssertSupportedVersion(ctx grpctx.Context, ver *csipb.Version) error
+	NodePublishVolume(
+		ctx grpctx.Context,
+		volumeid string,
+		readOnly bool,
+		targetPath string,
+		accessMode api.PersistentVolumeAccessMode,
+		fsType string,
+	) error
+	NodeUnpublishVolume(ctx grpctx.Context, volId string, targetPath string) error
+}
+
+// csiClient encapsulates all csi-plugin methods
+type csiDriverClient struct {
+	network    string
+	addr       string
+	conn       *grpc.ClientConn
+	idClient   csipb.IdentityClient
+	nodeClient csipb.NodeClient
+}
+
+func newCsiDriverClient(network, addr string) *csiDriverClient {
+	return &csiDriverClient{network: network, addr: addr}
+}
+
+// assertConnection ensures a valid connection has been established
+// if not, it creates a new connection and associated clients
+func (c *csiDriverClient) assertConnection() error {
+	if c.conn == nil {
+		conn, err := grpc.Dial(
+			c.addr,
+			grpc.WithInsecure(),
+			grpc.WithDialer(func(target string, timeout time.Duration) (net.Conn, error) {
+				return net.Dial(c.network, target)
+			}),
+		)
+		if err != nil {
+			return err
+		}
+		c.conn = conn
+		c.idClient = csipb.NewIdentityClient(conn)
+		c.nodeClient = csipb.NewNodeClient(conn)
+	}
+
+	return nil
+}
+
+// AssertCSIVersion determines if driver supports specified version
+func (c *csiDriverClient) AssertSupportedVersion(ctx grpctx.Context, ver *csipb.Version) error {
+	if err := c.assertConnection(); err != nil {
+		return err
+	}
+
+	rsp, err := c.idClient.GetSupportedVersions(ctx, &csipb.GetSupportedVersionsRequest{})
+	if err != nil {
+		return err
+	}
+	supported := false
+	vers := rsp.GetSupportedVersions()
+	for _, v := range vers {
```

Comment:
I'm confused on the contract between the caller and the plugin from this call. Does it mean that if, for example, the client is at v1.2.0, and the driver is at v0.9.0, but the client support v0.9.0, that the client should switch to that version? What does it mean to support different versions?

This is not really a question for just @vladimirvivien , but to the entire team: @saad-ali @jsafrane @childsb

Reply:
@lpabon not sure what the exact behavior is for multiple versions.